File: apps/mobile/lib/data/repositories/class_repository_impl.dart
import '../../core/error/failures.dart';
import '../../core/utils/either.dart';
import '../../domain/entities/class/class_entity.dart';
import '../../domain/entities/class/class_selector_entity.dart';
import '../../domain/entities/query/default_query_entity.dart';
import '../../domain/repositories/class_repository.dart';
import '../models/school/class_model.dart';
import '../datasources/remote/school/class_remote_data_source.dart';

class ClassRepositoryImpl implements ClassRepository {
  final ClassRemoteDataSource remoteDataSource;

  ClassRepositoryImpl({required this.remoteDataSource});

  @override
  Future<Either<Failure, List<ClassSelectorEntity>>> getClassesBySchool({
    required String schoolId,
    String? grade,
  }) async {
    try {
      final classes = await remoteDataSource.getClassesBySchool(
        schoolId: schoolId,
        grade: grade,
      );
      return Right(classes);
    } on Failure catch (e) {
      return Left(e);
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }

  @override
  Future<Either<Failure, List<ClassModel>>> getAllClasses(
    DefaultQueryEntity query,
  ) async {
    try {
      final classes = await remoteDataSource.getAllClasses(query);
      return Right(classes);
    } on Failure catch (e) {
      return Left(e);
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }

  @override
  Future<Either<Failure, ClassModel>> createClass(
    ClassEntity createClassData,
  ) async {
    try {
      final model = ClassModel(
        id: createClassData.id,
        name: createClassData.name,
        code: createClassData.code,
        schoolId: createClassData.schoolId,
        grade: createClassData.grade,
        level: createClassData.level,
        maxStudents: createClassData.maxStudents,
        currentStudents: createClassData.currentStudents,
        createdAt: createClassData.createdAt,
        updatedAt: createClassData.updatedAt,
      );
      final result = await remoteDataSource.createClass(model);
      return Right(result);
    } on Failure catch (e) {
      return Left(e);
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }

  @override
  Future<Either<Failure, ClassModel>> getClassById(String id) async {
    try {
      final result = await remoteDataSource.getClassById(id);
      return Right(result);
    } on Failure catch (e) {
      return Left(e);
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }

  @override
  Future<Either<Failure, ClassModel>> updateClass(
    ClassEntity updateClassData,
  ) async {
    try {
      final model = ClassModel(
        id: updateClassData.id,
        name: updateClassData.name,
        code: updateClassData.code,
        schoolId: updateClassData.schoolId,
        grade: updateClassData.grade,
        level: updateClassData.level,
        maxStudents: updateClassData.maxStudents,
        currentStudents: updateClassData.currentStudents,
        createdAt: updateClassData.createdAt,
        updatedAt: updateClassData.updatedAt,
      );
      final result = await remoteDataSource.updateClass(model);
      return Right(result);
    } on Failure catch (e) {
      return Left(e);
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }

  @override
  Future<Either<Failure, void>> deleteClass(String id) async {
    try {
      await remoteDataSource.deleteClass(id);
      return const Right(null);
    } on Failure catch (e) {
      return Left(e);
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }
}

File: services/user_service/src/data/repositories/ClassRepositoryImpl.ts
import {
  IClassRepository,
  ClassFilterOptions,
  SearchClassesOptions,
} from "../../domain/repositories/IClassRepository";
import { Class } from "../../domain/entities/Class";
import { ClassModel } from "../datasources/postgres/models/ClassModel";
import { GradeModel } from "../datasources/postgres/models/GradeModel";
import { Op } from "sequelize";

export class ClassRepositoryImpl implements IClassRepository {
  async create(classEntity: Class): Promise<Class> {
    const model = await ClassModel.create({
      name: classEntity.name,
      code: classEntity.code,
      schoolId: classEntity.schoolId,
      gradeId: classEntity.gradeId,
      maxStudents: classEntity.maxStudents,
      baseFee: classEntity.baseFee,
    });
    return this.toEntity(model);
  }

  async findById(id: string): Promise<Class | null> {
    const model = await ClassModel.findByPk(id);
    if (!model) return null;
    return this.toEntity(model);
  }

  async findByCode(code: string): Promise<Class | null> {
    const model = await ClassModel.findOne({ where: { code } });
    if (!model) return null;
    return this.toEntity(model);
  }

  async findBySchoolId(schoolId: string): Promise<Class[]> {
    const models = await ClassModel.findAll({ where: { schoolId } });
    return models.map((model) => this.toEntity(model));
  }

  async update(classEntity: Class): Promise<Class> {
    const [affectedCount, updatedModels] = await ClassModel.update(
      {
        name: classEntity.name,
        code: classEntity.code,
        schoolId: classEntity.schoolId,
        gradeId: classEntity.gradeId,
        maxStudents: classEntity.maxStudents,
        baseFee: classEntity.baseFee,
      },
      {
        where: { id: classEntity.id },
        returning: true,
      }
    );

    if (affectedCount === 0 || !updatedModels[0]) {
      throw new Error("Class not found or not updated");
    }

    return this.toEntity(updatedModels[0]);
  }

  async delete(id: string): Promise<boolean> {
    const deletedCount = await ClassModel.destroy({ where: { id } });
    return deletedCount > 0;
  }

  async getClassesBySchool(params: {
    schoolId: string;
    grade?: string;
  }): Promise<any[]> {
    const { SchoolModel } = await import(
      "../datasources/postgres/models/SchoolModel"
    );

    const whereCondition: any = {
      schoolId: params.schoolId,
      deletedAt: null,
    };

    // Add grade filter if provided
    if (params.grade) {
      whereCondition.gradeId = params.grade;
    }

    const classes = await ClassModel.findAll({
      where: whereCondition,
      include: [
        {
          model: SchoolModel,
          as: "school",
          attributes: ["id", "name", "level"],
        },
      ],
      order: [["name", "ASC"]],
    });

    return classes.map((cls) => {
      const clsAny = cls as any;
      return {
        id: cls.id,
        name: cls.name,
        code: cls.code,
        schoolId: cls.schoolId,
        grade: cls.gradeId,
        level: clsAny.school?.level || "",
        maxStudents: cls.maxStudents,
        currentStudents: clsAny.currentStudents || 0,
        createdAt: cls.createdAt,
        updatedAt: cls.updatedAt,
      };
    });
  }

  async findAllWithPagination(
    skip: number,
    limit: number,
    sortBy: string,
    sortOrder: "asc" | "desc",
    filters: ClassFilterOptions
  ): Promise<Class[]> {
    const whereClause: any = {};

    if (filters.schoolId) whereClause.schoolId = filters.schoolId;
    if (filters.gradeId) whereClause.gradeId = filters.gradeId;
    if (filters.maxStudents) whereClause.maxStudents = filters.maxStudents;
    if (filters.active !== undefined) whereClause.active = filters.active;

    const models = await ClassModel.findAll({
      where: whereClause,
      offset: skip,
      limit: limit,
      order: [[sortBy, sortOrder.toUpperCase() as "ASC" | "DESC"]],
    });

    return models.map((model) => this.toEntity(model));
  }

  async count(filters: ClassFilterOptions): Promise<number> {
    const whereClause: any = {};

    if (filters.schoolId) whereClause.schoolId = filters.schoolId;
    if (filters.gradeId) whereClause.gradeId = filters.gradeId;
    if (filters.maxStudents) whereClause.maxStudents = filters.maxStudents;
    if (filters.active !== undefined) whereClause.active = filters.active;

    return await ClassModel.count({ where: whereClause });
  }

  async searchClasses(schoolId: string): Promise<Class[]> {
    const whereClause: any = {};

    if (schoolId) whereClause.schoolId = schoolId;

    const models = await ClassModel.findAll({
      where: whereClause,
      order: [["name", "ASC"]],
      include: [
        {
          model: GradeModel,
          as: "grade",
          attributes: ["gradeGroup"],
        },
      ],
    });

    return models.map((model) => {
      const entity = this.toEntity(model);
      if (model.grade) {
        entity.gradeGroup = model.grade.gradeGroup;
      }
      return entity;
    });
  }

  private toEntity(model: ClassModel): Class {
    // Model has GradeModel included as 'grade' via association?
    // Note: ClassModel definition might need 'grade' property for typing if not using 'any' cast
    // For now we handle gradeGroup mapping in the calling method if needed or extend here

    return new Class(
      model.id,
      model.name,
      model.code,
      model.schoolId,
      model.gradeId,
      model.maxStudents,
      model.baseFee,
      [], // Students list not loaded by default in simple CRUD
      model.createdAt,
      model.updatedAt
    );
  }
}
